frontend/inst_pointer_chase_gen.py

```diff
@@ -0,0 +1,118 @@
+NO_CALLEE = -1
+
+
+class InstPointerChaseGenerator(common.BaseGenerator):
+    """Generates an instruction pointer chase benchmark."""
+
+    def __init__(self, depth, num_callchains, function_selector=None):
+        super().__init__()
+        self._depth = depth
+        self._num_callchains = num_callchains
+        self._caller2callee = {}
+        self._callchain_entry_functions = []
+        self._entry_function_id = None
+        if function_selector is None:
+            self._function_selector = common.PopRandomElement
+        else:
+            self._function_selector = function_selector
+        self._function_body = self._AddCodeBlockBody('int x = 1;\n'
+                                                     'int y = x*x + 3;\n'
+                                                     'int z = y*x + 12345;\n'
+                                                     'int w = z*z + x - y;\n')
+
+    def _GenerateCallchainMappings(self):
+        num_functions = self._num_callchains * self._depth
+        function_list = list(range(0, num_functions))
+        for _ in range(0, self._num_callchains):
+            # Generate a pointer chase. The final function will just return.
+            caller = self._function_selector(function_list)
+            self._callchain_entry_functions.append(caller)
+            for _ in range(0, self._depth - 1):
+                callee = self._function_selector(function_list)
+                self._caller2callee[caller] = callee
+                caller = callee
+            # The final function in the chain does not call.
+            self._caller2callee[caller] = NO_CALLEE
+
+        assert len(self._caller2callee) == num_functions, (
+            'There should be exactly one caller2callee mapping for every '
+            'function.')
+
+    def _GenerateCallchainFunctions(self):
+        # First, generate codeblocks. Each function has two: the main body, with
+        # a fallthrough branch, and the call, with a return terminator branch.
+
+        for caller, callee in self._caller2callee.items():
+            function = self._AddFunctionWithId(caller)
+            main_body = self._AddCodeBlock()
+            main_body.code_block_body_id = self._function_body.id
+            main_body.terminator_branch.type = \
+                cfg_pb2.Branch.BranchType.FALLTHROUGH
+            function.instructions.append(main_body)
+
+            if callee != NO_CALLEE:
+                call_block = self._AddCodeBlock()
+                # Leave the branch target unspecified for now.
+                call_block.terminator_branch.type = (
+                    cfg_pb2.Branch.BranchType.DIRECT_CALL)
+                call_block.terminator_branch.targets.append(callee)
+                call_block.terminator_branch.taken_probability.append(1)
+                # The end of a function in C will implicitly return, no need to
+                # create another CodeBlock.
+                function.instructions.append(call_block)
```

Comment:
Duplicate code in lines 93-97, 69-72

Extract function. Possibly to common class.

Reply:
The code looks rather different to me actually. One is for a fallthrough, while the other is for a direct call, which requires a target. Doesn't seem worthwhile to extract.